### PR TITLE
Split Phabricator and Mercurial actions

### PR DESCRIPTION
**Pull request by La0**
Last big part before splitting pulselistener in lib & client applications. Refs #2185.

This PR changes several important things:

- `MercurialWorker` only applies `PhabricatorBuild` and their pre-loaded stack of patches: there is no Phabricator API calls there anymore.
- A new generic `PhabricatorActions` class is available in the library: it has the only `PhabricatorAPI` instance in pulselistener, and exposes it through high level functions. No queue usage here, as it's the responsability of the client application.
- `PhabricatorBuild` does not have any inner logic anymore, it simply stores data.
- A `PhabricatorPatch` tuple is added to store efficiently the stack of patches, which is built only from Phabricator (no repository check when creating the stack)
- A `PhabricatorCodeReview` class extending `PhabricatorActions` is added in the client application, and contains the workflow specific to code review: building specific mercurial repositories, pre-loading builds with stack of patches & reviewers once they reach the public state.
- A new queue is added in the application, to pre-load Phabricator builds with extended data *before* reaching the mercurial worker.

In the end, the code review workflow through queues is:
1. Build triggered through the webserver and stored in `QUEUE_WEB_BUILDS`
2. `QUEUE_WEB_BUILDS` is read by `PhabricatorCodeReview`, once the build is seen as public, it is pre-loaded with Phabricator data, and pused in `QUEUE_CODE_REVIEW`.
3. `QUEUE_CODE_REVIEW` is read by `HookPhabricator` which just pushes it to `QUEUE_MERCURIAL` & `QUEUE_PHABRICATOR_RESULTS` (+ risk analysis workflow)

### CODE REVIEW

File: src/pulselistener/pulselistener/code_review.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    '''
+    def __init__(self, publish=False, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.publish = publish
+        logger.info('Phabricator publication is {}'.format(self.publish and 'enabled' or 'disabled'))
+
+    def register(self, bus):
+        self.bus = bus
+
+    def build_repositories(self, repositories, cache_root):
```
Comment (marco-c) [inline src/pulselistener/pulselistener/code_review.py]: Nit: `get_repositories`?

File: src/pulselistener/pulselistener/listener.py [hunk 2]
```diff
@@ -33,26 +34,6 @@ def __init__(self, configuration, bus):
         assert 'hookId' in configuration
```
Comment (marco-c) [inline src/pulselistener/pulselistener/listener.py]: I was thinking, what if we merge HookPhabricator and PhabricatorCodeReview?

File: src/pulselistener/pulselistener/listener.py [hunk 3]
```diff
@@ -280,28 +247,17 @@ def __init__(self,
...
+        # Run client applications
         consumers = [
-            hook.run()
-            for hook in hooks
+            client.run()
+            for client in self.clients
         ]
 
+        # Add build details loading step
+        consumers.append(self.phabricator.load_builds(QUEUE_WEB_BUILDS, QUEUE_CODE_REVIEW))
```
Comment (marco-c) [inline src/pulselistener/pulselistener/listener.py]: Could we move this directly into PhabricatorCodeReview? Maybe after we get rid of pulselistener

File: src/pulselistener/pulselistener/lib/phabricator.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        self.revision = None
+        self.reviewers = []
+        self.diff = None
+        self.stack = []
+
+    def __str__(self):
+        return 'Revison {} - {}'.format(self.revision_id, self.target_phid)
+
+
+class PhabricatorActions(object):
```
Comment (marco-c) [inline src/pulselistener/pulselistener/lib/phabricator.py]: We could merge this too into PhabricatorCodeReview (it's the only user). Or were you thinking something else could be using this in the future?

File: src/pulselistener/pulselistener/lib/phabricator.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+            projects = set(build.revision['attachments']['projects']['projectPHIDs'])
+            if projects.intersection(self.secure_projects):
+                raise Exception('Secure revision')
+        except Exception as e:
+            logger.info('Revision not accessible', build=str(build), error=str(e))
+            return False
+
+        return True
+
+    def load_patches_stack(self, build):
```
Comment (marco-c) [inline src/pulselistener/pulselistener/lib/phabricator.py]: I'd keep this into libmozdata, and remove the `revision_available` stuff from there (or make it optional) so you are not forced to use it.

File: src/pulselistener/pulselistener/lib/mercurial.py [hunk 6]
```diff
@@ -15,9 +15,11 @@
 
 import hglib
 import structlog
+from libmozdata.phabricator import revision_available
```
Comment (marco-c) [inline src/pulselistener/pulselistener/lib/mercurial.py]: I'd reimplement `revision_available` here (it's just a few lines), so we don't have to import phabricator here

File: src/pulselistener/pulselistener/lib/phabricator.py (hunk 4 above)
Comment (La0) [reply src/pulselistener/pulselistener/lib/phabricator.py]: Yes, i wanted to have these shared tool in the library, thinking that we would end up with a code review client & a risk analysis client.

**Pull request merged**